### ISSUE TITLE
IndexOutOfBoundsException when calling PostgreSQL table valued function returning empty result, with <tableValuedFunction/> turned off

### ISSUE THREAD

**Issue by meiermarkus** [jOOQ/jOOQ#5844]
This is a bit a corner case, but when a Postgres function (that returns a TABLE) doesn't return anything, jooq will throw an `IndexOutOfBoundsException`.
This seems to be caused in `executeSelectFromPOSTGRES `on line 400 (in master) since `result `isn't checked before accessing.

```
java.lang.IndexOutOfBoundsException: Index: 0, Size: 0
	at java.util.ArrayList.rangeCheck(ArrayList.java:653)
	at java.util.ArrayList.get(ArrayList.java:429)
	at org.jooq.impl.ResultImpl.get(ResultImpl.java:2420)
	at org.jooq.impl.ResultImpl.getValue(ResultImpl.java:300)
	at org.jooq.impl.AbstractRoutine.executeSelectFromPOSTGRES(AbstractRoutine.java:398)
	at org.jooq.impl.AbstractRoutine.execute(AbstractRoutine.java:293)
	at org.jooq.impl.AbstractRoutine.execute(AbstractRoutine.java:275)
```

**Comment by lukaseder** [comment]
Thank you very much for your report. Unfortunately, I cannot reproduce it yet, from your description. I've tried this function here:

```sql
CREATE OR REPLACE FUNCTION f5844()
RETURNS TABLE (
  i INTEGER
)
AS $$
BEGIN
END;
$$ LANGUAGE plpgsql;
```

And this test works:

```java
    @Test
    public void testPostgresTableValuedFunctionWithoutResults() {
        Result<F5844Record> r0 = F5844(create().configuration());
        Result<F5844Record> r1 = create().selectFrom(F5844()).fetch();
        Result<F5844Record> r2 = Routines.f5844(create().configuration());
        assertEquals(0, r0.size());
        assertEquals(0, r1.size());
        assertEquals(0, r2.size());
    }
```

But from your line numbers, I'm assuming:

1) either you have out parameters or something else that I'm missing
2) you perhaps deactivated the `<tableValuedFunctions/>` flag in the code generator?

**Comment by meiermarkus** [comment]
Here's the signature of the function:
```
CREATE FUNCTION nearest_network(lat numeric, lon numeric, max_distance_m double precision) RETURNS TABLE(network_id integer, distance double precision)
    LANGUAGE plpgsql
    SET default_tablespace TO ''
    AS $$
BEGIN
```

And this is the java code using it (throws on the `.execute()` line):
```
        NearestNetwork nn = new NearestNetwork();
        nn.setLat(BigDecimal.valueOf(latitude));
        nn.setLon(BigDecimal.valueOf(longitude));
        nn.setMaxDistanceM(maxDistanceInMeters);
        nn.execute(getContext().configuration());
        networkId = nn.getNetworkId();
```

**Comment by lukaseder** [comment]
Can you confirm you're using `<tableValuedFunctions>false</tableValuedFunctions>` in your code generator?

**Comment by meiermarkus** [comment]
We didn't set this option in the configuration. Currently we're using 3.8.6 version, the option seems unavailable.
Using the latest version, and `tableValuedFunctions` enabled, things seem to work! Disabling `tableValuedFunctions` causes that behavior, as you've expected.

**Comment by lukaseder** [comment]
Hmm, interesting. It should be available in jOOQ 3.8 (https://www.jooq.org/xsd/jooq-codegen-3.8.0.xsd). Anyway, I remember now. Also, we fixed this issue here https://github.com/jOOQ/jOOQ/issues/5349 in jOOQ 3.8.4. Perhaps, you had a mismatch between jOOQ versions of your runtime and your code generator?

Anyway, the routine call should work regardless if the flag is activated or not. Thanks for the pointers.

**Comment by meiermarkus** [comment]
Yeah, I just noticed that the option is available on 3.8 as well. When enabling it and adjusting the code, it works as well.
There shouldn't be a mismatch with runtime and generated code...

Thanks for your help as well!

**Comment by lukaseder** [comment]
Fixed in versions:

- 3.20.0
- 3.19.15 (https://github.com/jOOQ/jOOQ/issues/17467)
- 3.18.22 (https://github.com/jOOQ/jOOQ/issues/17468)
- 3.17.31 (https://github.com/jOOQ/jOOQ/issues/17469)
